libplatsupport/plat_include/zynqmp/platsupport/plat/timer.h

```diff
@@ -0,0 +1,144 @@
+
+#define TTC_TIMER_SIZE           0x1000
+#define TTC0_TIMER_SIZE          TTC_TIMER_SIZE
+#define TTC1_TIMER_SIZE          TTC_TIMER_SIZE
+#define TTC2_TIMER_SIZE          TTC_TIMER_SIZE
+#define TTC3_TIMER_SIZE          TTC_TIMER_SIZE
+
+/* IRQs */
+#define TTC0_TIMER1_IRQ          36
+#define TTC0_TIMER2_IRQ          37
+#define TTC0_TIMER3_IRQ          38
+#define TTC1_TIMER1_IRQ          39
+#define TTC1_TIMER2_IRQ          40
+#define TTC1_TIMER3_IRQ          41
+#define TTC2_TIMER1_IRQ          42
+#define TTC2_TIMER2_IRQ          43
+#define TTC2_TIMER3_IRQ          44
+#define TTC3_TIMER1_IRQ          45
+#define TTC3_TIMER2_IRQ          46
+#define TTC3_TIMER3_IRQ          47
+
+/* Timers */
+typedef enum timer_id {
+    TTC0_TIMER1,
+    TTC0_TIMER2,
+    TTC0_TIMER3,
+    TTC1_TIMER1,
+    TTC1_TIMER2,
+    TTC1_TIMER3,
+    TTC2_TIMER1,
+    TTC2_TIMER2,
+    TTC2_TIMER3,
+    TTC3_TIMER1,
+    TTC3_TIMER2,
+    TTC3_TIMER3,
+    NTIMERS
+} ttc_id_t;
+#define TMR_DEFAULT TTC0_TIMER1
+
+static const uintptr_t zynq_timer_paddrs[] = {
+    [TTC0_TIMER1] = TTC0_PADDR,
+    [TTC0_TIMER2] = TTC0_PADDR,
+    [TTC0_TIMER3] = TTC0_PADDR,
+    [TTC1_TIMER1] = TTC1_PADDR,
+    [TTC1_TIMER2] = TTC1_PADDR,
+    [TTC1_TIMER3] = TTC1_PADDR,
+    [TTC2_TIMER1] = TTC2_PADDR,
+    [TTC2_TIMER2] = TTC2_PADDR,
+    [TTC2_TIMER3] = TTC2_PADDR,
+    [TTC3_TIMER1] = TTC3_PADDR,
+    [TTC3_TIMER2] = TTC3_PADDR,
+    [TTC3_TIMER3] = TTC3_PADDR
+};
+
+static const int zynq_timer_irqs[] = {
+    [TTC0_TIMER1] = TTC0_TIMER1_IRQ,
+    [TTC0_TIMER2] = TTC0_TIMER2_IRQ,
+    [TTC0_TIMER3] = TTC0_TIMER3_IRQ,
+    [TTC1_TIMER1] = TTC1_TIMER1_IRQ,
+    [TTC1_TIMER2] = TTC1_TIMER2_IRQ,
+    [TTC1_TIMER3] = TTC1_TIMER3_IRQ,
+    [TTC2_TIMER1] = TTC2_TIMER1_IRQ,
+    [TTC2_TIMER2] = TTC2_TIMER2_IRQ,
+    [TTC2_TIMER3] = TTC2_TIMER3_IRQ,
+    [TTC3_TIMER1] = TTC3_TIMER1_IRQ,
+    [TTC3_TIMER2] = TTC3_TIMER2_IRQ,
+    [TTC3_TIMER3] = TTC3_TIMER3_IRQ
+};
+
+typedef struct {
+    /* vaddr pwm is mapped to */
+    void *vaddr;
+    clk_t* clk_src;
+    ttc_id_t id;
+} ttc_config_t;
+
+typedef struct {
+    void *regs;
+    clk_t clk;
+    freq_t freq;
+    ttc_id_t id;
+} ttc_t;
+
+static UNUSED timer_properties_t ttc_properties = {
+    .upcounter = true,
+    .timeouts = true,
+    .bit_width = 16,
+    .irqs = 1,
+    .relative_timeouts = true,
+    .absolute_timeouts = true
+};
+
+static inline uintptr_t ttc_paddr(ttc_id_t id)
+{
+    if (id >= TTC0_TIMER1 && id < NTIMERS) {
+        return zynq_timer_paddrs[id];
+    } else {
+        return 0;
+    }
+}
+
+static inline int ttc_irq(ttc_id_t id)
+{
+    if (id >= TTC0_TIMER1 && id < NTIMERS) {
+        return zynq_timer_irqs[id];
+    } else {
+        return 0;
+    }
+}
+
+int ttc_init(ttc_t *ttc, ttc_config_t config);
+int ttc_start(ttc_t *ttc);
+int ttc_stop(ttc_t *ttc);
+int ttc_set_timeout(ttc_t *ttc, uint64_t ns, bool periodic);
+void ttc_handle_irq(ttc_t *ttc);
+uint64_t ttc_get_time(ttc_t *ttc);
```

Comment:
bit of a nitpick, but given that these are an interface to 'mach' code, then the interface should be moved to a 'mach' header and not duplicated across the two zynq platforms.

Reply:
I have a fix for this that I will be pushing up next week.